From 3aafd87c5cc85cbc0e9bc1fe25fe4a35fc813f6f Mon Sep 17 00:00:00 2001
From: Georg Neis <neis@chromium.org>
Date: Wed, 9 Jul 2025 02:14:18 +0000
Subject: [PATCH] Revert "base/hash: remove streaming md5 interface"

This reverts commit 4553ef7771c0c2d6b330b6e645a72e742057ea3b, which
removed some md5 functions still in use by platform2/metrics. Long-term
because closely tied to
long-term-0600-Use-non-boringssl-md5-sha1-implementation.patch.
---
 base/hash/md5.cc          | 16 +++++++++++
 base/hash/md5.h           | 38 ++++++++++++++++++++++++-
 base/hash/md5_unittest.cc | 59 +++++++++++++++++++++++++++++++++++++++
 3 files changed, 112 insertions(+), 1 deletion(-)

diff --git a/base/hash/md5.cc b/base/hash/md5.cc
index f799898ec0..7d68475a27 100644
--- a/base/hash/md5.cc
+++ b/base/hash/md5.cc
@@ -15,6 +15,22 @@
 #include "third_party/boringssl/src/include/openssl/md5.h"
 
 namespace base {
+void MD5Init(MD5Context* context) {
+  MD5_Init(context);
+}
+
+void MD5Update(MD5Context* context, std::string_view data) {
+  MD5Update(context, base::as_byte_span(data));
+}
+
+void MD5Update(MD5Context* context, base::span<const uint8_t> data) {
+  MD5_Update(context, data.data(), data.size());
+}
+
+void MD5Final(MD5Digest* digest, MD5Context* context) {
+  MD5_Final(digest->a.data(), context);
+}
+
 std::string MD5DigestToBase16(const MD5Digest& digest) {
   return ToLowerASCII(HexEncode(digest.a));
 }
diff --git a/base/hash/md5.h b/base/hash/md5.h
index 5add382a7a..c9050eb9c2 100644
--- a/base/hash/md5.h
+++ b/base/hash/md5.h
@@ -13,6 +13,8 @@
 
 #include "base/base_export.h"
 #include "base/containers/span.h"
+#include "build/build_config.h"
+#include "third_party/boringssl/src/include/openssl/md5.h"
 
 // MD5 stands for Message Digest algorithm 5.
 //
@@ -29,13 +31,47 @@
 // NEW USES OF THIS API ARE FORBIDDEN FOR ANY PURPOSE. INSTEAD, YOU MUST USE
 // //crypto/obsolete/md5.h.
 
+// These functions perform MD5 operations. The simplest call is MD5Sum() to
+// generate the MD5 sum of the given data.
+//
+// You can also compute the MD5 sum of data incrementally by making multiple
+// calls to MD5Update():
+//   MD5Context ctx; // intermediate MD5 data: do not use
+//   MD5Init(&ctx);
+//   MD5Update(&ctx, data1, length1);
+//   MD5Update(&ctx, data2, length2);
+//   ...
+//
+//   MD5Digest digest; // the result of the computation
+//   MD5Final(&digest, &ctx);
+//
+// You can call MD5DigestToBase16() to generate a string of the digest.
+
 namespace base {
 
 // The output of an MD5 operation.
 struct MD5Digest {
-  std::array<uint8_t, 16> a;
+  std::array<uint8_t, MD5_DIGEST_LENGTH> a;
 };
 
+// Used for storing intermediate data during an MD5 computation. Callers
+// should not access the data.
+using MD5Context = MD5_CTX;
+
+// Initializes the given MD5 context structure for subsequent calls to
+// MD5Update().
+BASE_EXPORT void MD5Init(MD5Context* context);
+
+// For the given buffer of |data| as a std::string_view or span, updates the
+// given MD5 context with the sum of the data. You can call this any number of
+// times during the computation, except that MD5Init() must have been called
+// first.
+BASE_EXPORT void MD5Update(MD5Context* context, std::string_view data);
+BASE_EXPORT void MD5Update(MD5Context* context, base::span<const uint8_t> data);
+
+// Finalizes the MD5 operation and fills the buffer with the digest.
+BASE_EXPORT void MD5Final(MD5Digest* digest, MD5Context* context);
+
 // Converts a digest into human-readable hexadecimal.
 BASE_EXPORT std::string MD5DigestToBase16(const MD5Digest& digest);
 
diff --git a/base/hash/md5_unittest.cc b/base/hash/md5_unittest.cc
index 4390fd8a5d..73590a6f9e 100644
--- a/base/hash/md5_unittest.cc
+++ b/base/hash/md5_unittest.cc
@@ -57,6 +57,50 @@ TEST(MD5, MD5SumLongData) {
   EXPECT_EQ(base::span(expected), digest.a);
 }
 
+TEST(MD5, ContextWithEmptyData) {
+  MD5Context ctx;
+  MD5Init(&ctx);
+
+  MD5Digest digest;
+  MD5Final(&digest, &ctx);
+
+  const uint8_t expected[] = {0xd4, 0x1d, 0x8c, 0xd9, 0x8f, 0x00, 0xb2, 0x04,
+                              0xe9, 0x80, 0x09, 0x98, 0xec, 0xf8, 0x42, 0x7e};
+  EXPECT_EQ(base::span(expected), digest.a);
+}
+
+TEST(MD5, ContextWithLongData) {
+  MD5Context ctx;
+  MD5Init(&ctx);
+
+  const size_t length = 10 * 1024 * 1024 + 1;
+  auto data = base::HeapArray<uint8_t>::Uninit(length);
+
+  size_t i = 0;
+  for (auto& datum : data) {
+    datum = i++ & 0xff;
+  }
+
+  size_t total = 0;
+  while (total < data.size()) {
+    size_t len = 4097;  // intentionally not 2^k.
+    if (len > length - total) {
+      len = length - total;
+    }
+
+    MD5Update(&ctx, data.subspan(total, len));
+    total += len;
+  }
+
+  EXPECT_EQ(length, total);
+
+  MD5Digest digest;
+  MD5Final(&digest, &ctx);
+
+  const uint8_t expected[] = {0x90, 0xbd, 0x6a, 0xd9, 0x0a, 0xce, 0xf5, 0xad,
+                              0xaa, 0x92, 0x20, 0x3e, 0x21, 0xc7, 0xa1, 0x3e};
+  EXPECT_EQ(base::span(expected), digest.a);
+}
 
 // Example data from http://www.ietf.org/rfc/rfc1321.txt A.5 Test Suite
 TEST(MD5, MD5StringTestSuite1) {
@@ -108,4 +152,19 @@ TEST(MD5, MD5StringTestSuite7) {
   EXPECT_EQ(expected, actual);
 }
 
+TEST(MD5, ContextWithStringData) {
+  MD5Context ctx;
+  MD5Init(&ctx);
+
+  MD5Update(&ctx, "abc");
+
+  MD5Digest digest;
+  MD5Final(&digest, &ctx);
+
+  std::string actual = MD5DigestToBase16(digest);
+  std::string expected = "900150983cd24fb0d6963f7d28e17f72";
+
+  EXPECT_EQ(expected, actual);
+}
+
 }  // namespace base
-- 
2.50.0.727.gbf7dc18ff4-goog

